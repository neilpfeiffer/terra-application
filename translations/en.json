{
  "terraApplication.unsavedChangesPrompt.title": "Unsaved Changes",
  "terraApplication.unsavedChangesPrompt.singlePromptMessage": "Unsaved changes were made to {promptDescription}.",
  "terraApplication.unsavedChangesPrompt.multiplePromptMessageIntro": "Unsaved changes were made to the following:",
  "terraApplication.unsavedChangesPrompt.multiplePromptMessageOutro": "Changes will be lost if you don't save them. How do you want to proceed?",
  "terraApplication.unsavedChangesPrompt.acceptButtonText": "Don't Save",
  "terraApplication.unsavedChangesPrompt.rejectButtonText": "Continue Editing",
  "terraApplication.errorBoundary.defaultErrorMessage": "The system encountered an error: {errorDetails}",
  "Terra.applicationNavigation.tabs.rollupButtonDescription": "More Navigation",
  "Terra.applicationNavigation.tabs.rollupMenuHeaderTitle": "More Navigation",
  "Terra.applicationNavigation.header.utilityButtonTitleUser": "User Settings",
  "Terra.applicationNavigation.header.utilityButtonTitleNoUser": "Options",
  "Terra.applicationNavigation.drawerMenu.navigation": "Navigation Items",
  "Terra.applicationNavigation.drawerMenu.utilities": "Utility Items",
  "Terra.applicationNavigation.utilityMenu.logout": "Logout",
  "Terra.applicationNavigation.utilityMenu.settings": "Settings",
  "Terra.applicationNavigation.utilityMenu.help": "Help",
  "Terra.applicationNavigation.utilityMenu.headerTitle": "Utilities",
  "Terra.applicationNavigation.extensions.rollupButtonDescription": "Additional Tools",
  "Terra.applicationNavigation.extensions.rollupMenuHeaderTitle": "Additional Tools",
  "Terra.applicationNavigation.tabs.rollupButtonTitle": "More",
  "Terra.applicationNavigation.notifications.new": "New",
  "Terra.applicationNavigation.header.skipToContentTitle": "Skip to Content",
  "Terra.applicationNavigation.header.menuButtonTitle": "Menu"
}
