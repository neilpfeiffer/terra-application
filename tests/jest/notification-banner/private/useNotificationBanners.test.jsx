import React, { useState } from 'react';
import { act } from 'react-dom/test-utils';
import { render, fireEvent } from '@testing-library/react';
import { IntlProvider } from 'react-intl';

import BannerRegistrationContext from '../../../../src/notification-banner/private/BannerRegistrationContext';
import NotificationBanner from '../../../../src/notification-banner';
import useNotificationBanners from '../../../../src/notification-banner/private/useNotificationBanners';
// eslint-disable-next-line import/no-unresolved, import/extensions
import { messages } from '../../../../aggregated-translations/en'; // aggregation is pre-jest step so this will exist

global.console.warn = jest.fn();

let keyValue = 0;
const mockBannerProps = { variant: 'error' };

// eslint-disable-next-line react/prop-types
const ChildContent = ({ showBannerOnRender = false, buttonId = 'show-banner', children = null }) => {
  const [hasBanner, setHasBanner] = useState(showBannerOnRender);
  keyValue += 1;
  return (
    <div>
      <button
        type="button"
        data-testid={buttonId}
        onClick={() => setHasBanner(!hasBanner)}
      >
        Toggle Notification Banner
      </button>
      {hasBanner ? <NotificationBanner {...mockBannerProps} key={keyValue} /> : null}
      {children}
    </div>
  );
};

const renderComponentWithChild = (childrenContent) => {
  let childContext;

  const ExampleWithBannerProvider = () => {
    const { NotificationBannerProvider, NotificationBanners } = useNotificationBanners();

    return (
      <IntlProvider locale="en" messages={messages}>
        <NotificationBannerProvider>
          <NotificationBanners />
          <BannerRegistrationContext.Consumer>
            {(context) => {
              childContext = context;
              return childrenContent;
            }}
          </BannerRegistrationContext.Consumer>
        </NotificationBannerProvider>
      </IntlProvider>
    );
  };
  const component = render(<ExampleWithBannerProvider />);

  return { component, context: childContext };
};

describe('useNotificationBanners', () => {
  beforeEach(() => {
    jest.resetAllMocks;
    // eslint-disable-next-line no-console
    console.warn.mockClear();
  });

  it('renders with children', () => {
    const { component } = renderComponentWithChild(null);
    expect(component.container).toMatchSnapshot();
  });

  it('renders with children', () => {
    const { component } = renderComponentWithChild(<div>child</div>);

    expect(component.container).toMatchSnapshot();
  });

  it('renders with children that has a an banner', () => {
    const { component } = renderComponentWithChild(<ChildContent showBannerOnRender />);

    expect(component.container).toMatchSnapshot();
  });

  describe('NotificationBannerProvider.registerNotificationBanner', () => {
    it('provides registerNotificationBanner callback in context', () => {
      const { context } = renderComponentWithChild();

      expect(context).toBeDefined();
      expect(context).toHaveProperty('registerNotificationBanner', expect.any(Function));
    });

    it('does not register banner if no id is provided', () => {
      const { component, context } = renderComponentWithChild();
      expect(component.container.querySelectorAll('[data-terra-application-notification-banner]')).toHaveLength(0);

      expect(() => context.registerNotificationBanner(undefined, mockBannerProps)).toThrowError('A banner cannot be registered without an identifier.');

      expect(component.container.querySelectorAll('[data-terra-application-notification-banner]')).toHaveLength(0);
    });

    it('registers banner when id is provided', () => {
      expect.assertions(3);
      const { component, context } = renderComponentWithChild();
      expect(component.container.querySelectorAll('[data-terra-application-notification-banner]')).toHaveLength(0);

      expect(() => act(() => {
        context.registerNotificationBanner('mockID', mockBannerProps);
      })).not.toThrowError('A banner cannot be registered without an identifier.');

      expect(component.container.querySelectorAll('[data-terra-application-notification-banner]')).toHaveLength(1);
    });

    it('registers banner when rendered after mount', () => {
      const { component } = renderComponentWithChild(<ChildContent />);

      expect(component.container).toMatchSnapshot();

      const childButton = component.getByTestId('show-banner');
      act(() => {
        fireEvent.click(childButton);
      });

      expect(component.container.querySelectorAll('[data-terra-application-notification-banner]')).toHaveLength(1);
      expect(component.container).toMatchSnapshot();
    });
  });

  describe('NotificationBannerProvider.unregisterNotificationBanner', () => {
    it('provides unregisterNotificationBanner callback in context', () => {
      const { context } = renderComponentWithChild();

      expect(context).toBeDefined();
      expect(context).toHaveProperty('unregisterNotificationBanner', expect.any(Function));
    });

    it('does not unregister notification banner if no id is provided', () => {
      const { component, context } = renderComponentWithChild(<ChildContent showBannerOnRender />);

      // eslint-disable-next-line no-console
      console.warn.mockClear(); // clear terra-responsive-element will be deprecated warning

      expect(component.container.querySelectorAll('[data-terra-application-notification-banner]')).toHaveLength(1);

      expect(() => context.unregisterNotificationBanner(undefined)).toThrowError('A banner cannot be unregistered without an identifier or banner variant.');

      expect(component.container.querySelectorAll('[data-terra-application-notification-banner]')).toHaveLength(1);
    });

    it('does not unregister notification banner if banner variant is not provided', () => {
      const { component, context } = renderComponentWithChild(<ChildContent showBannerOnRender />);

      // eslint-disable-next-line no-console
      console.warn.mockClear(); // clear terra-responsive-element will be deprecated warning

      expect(component.container.querySelectorAll('[data-terra-application-notification-banner]')).toHaveLength(1);

      // context.unregisterNotificationBanner('mockID', undefined);
      expect(() => context.unregisterNotificationBanner('mockID', undefined)).toThrowError('A banner cannot be unregistered without an identifier or banner variant.');

      expect(component.container.querySelectorAll('[data-terra-application-notification-banner]')).toHaveLength(1);
    });

    it('unregisters notification banner when id is provided', () => {
      const { component } = renderComponentWithChild(<ChildContent showBannerOnRender />);
      expect(component.container.querySelectorAll('[data-terra-application-notification-banner]')).toHaveLength(1);

      expect(component.container).toMatchSnapshot();

      const childButton = component.getByTestId('show-banner');
      act(() => {
        fireEvent.click(childButton);
      });

      expect(component.container.querySelectorAll('[data-terra-application-notification-banner]')).toHaveLength(0);
      expect(component.container).toMatchSnapshot();
    });
  });

  describe(('Nested Notification Banner Provider'), () => {
    it('registers notification banner with top-level Banner Provider', () => {
      const ChildComponent = () => {
        const { NotificationBannerProvider, NotificationBanners } = useNotificationBanners();
        return (
          <ChildContent>
            <NotificationBannerProvider>
              <NotificationBanners />
              <ChildContent buttonId="show-2nd-banner" />
            </NotificationBannerProvider>
          </ChildContent>
        );
      };
      const { component } = renderComponentWithChild(<ChildComponent />);
      expect(component.container).toMatchSnapshot();

      // trigger top Provider's banner
      const childButton = component.getByTestId('show-banner');
      act(() => {
        fireEvent.click(childButton);
      });

      expect(component.container.querySelectorAll('[data-terra-application-notification-banner]')).toHaveLength(1);
      expect(component.container).toMatchSnapshot();

      // trigger nested banner Provider's banner
      const secondChildButton = component.getByTestId('show-2nd-banner');
      act(() => {
        fireEvent.click(secondChildButton);
      });

      expect(component.container.querySelectorAll('[data-terra-application-notification-banner]')).toHaveLength(2);
      expect(component.container).toMatchSnapshot();
    });

    it('unregisters banner with top-level Banner Provider', () => {
      const ChildComponent = () => {
        const { NotificationBannerProvider, NotificationBanners } = useNotificationBanners();
        return (
          <ChildContent showBannerOnRender>
            <NotificationBannerProvider>
              <NotificationBanners />
              <ChildContent buttonId="show-2nd-banner" showBannerOnRender />
            </NotificationBannerProvider>
          </ChildContent>
        );
      };
      const { component } = renderComponentWithChild(<ChildComponent />);

      expect(component.container.querySelectorAll('[data-terra-application-notification-banner]')).toHaveLength(2);

      expect(component.container).toMatchSnapshot();

      // remove top Provider's banner
      const childButton = component.getByTestId('show-banner');
      act(() => {
        fireEvent.click(childButton);
      });

      expect(component.container.querySelectorAll('[data-terra-application-notification-banner]')).toHaveLength(1);
      expect(component.container).toMatchSnapshot();

      // remove nested banner Provider's banner
      const secondChildButton = component.getByTestId('show-2nd-banner');
      act(() => {
        fireEvent.click(secondChildButton);
      });

      expect(component.container.querySelectorAll('[data-terra-application-notification-banner]')).toHaveLength(0);
      expect(component.container).toMatchSnapshot();
    });
  });
});
