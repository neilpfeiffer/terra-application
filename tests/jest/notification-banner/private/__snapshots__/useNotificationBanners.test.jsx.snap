// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useNotificationBanners Nested Notification Banner Provider registers notification banner with top-level Banner Provider 1`] = `
<div>
  <div>
    <button
      data-testid="show-banner"
      type="button"
    >
      Toggle Notification Banner
    </button>
    <div>
      <button
        data-testid="show-2nd-banner"
        type="button"
      >
        Toggle Notification Banner
      </button>
    </div>
  </div>
</div>
`;

exports[`useNotificationBanners Nested Notification Banner Provider registers notification banner with top-level Banner Provider 2`] = `
<div>
  <div
    aria-live="polite"
  >
    <div
      class="alert-base error narrow"
      data-terra-application-notification-banner="error"
      role="alert"
    >
      <div
        class="body body-std"
      >
        <span
          class="icon"
        >
          <svg
            aria-hidden="true"
            class="tui-Icon icon IconError"
            focusable="false"
            height="1em"
            viewBox="0 0 48 48"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M24 0c13.3 0 24 10.7 24 24S37.3 48 24 48 0 37.3 0 24C0 10.8 10.7 0 23.9 0h.1z"
              fill="#E50000"
            />
            <path
              d="M7 21.5h34v5H7v-5z"
              fill="#FFF"
            />
          </svg>
        </span>
        <div
          class="section"
        >
          <strong
            class="title"
          >
            Terra.alert.error
          </strong>
        </div>
      </div>
    </div>
  </div>
  <div>
    <button
      data-testid="show-banner"
      type="button"
    >
      Toggle Notification Banner
    </button>
    <div>
      <button
        data-testid="show-2nd-banner"
        type="button"
      >
        Toggle Notification Banner
      </button>
    </div>
  </div>
</div>
`;

exports[`useNotificationBanners Nested Notification Banner Provider registers notification banner with top-level Banner Provider 3`] = `
<div>
  <div
    aria-live="polite"
  >
    <div
      class="alert-base error narrow"
      data-terra-application-notification-banner="error"
      role="alert"
    >
      <div
        class="body body-std"
      >
        <span
          class="icon"
        >
          <svg
            aria-hidden="true"
            class="tui-Icon icon IconError"
            focusable="false"
            height="1em"
            viewBox="0 0 48 48"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M24 0c13.3 0 24 10.7 24 24S37.3 48 24 48 0 37.3 0 24C0 10.8 10.7 0 23.9 0h.1z"
              fill="#E50000"
            />
            <path
              d="M7 21.5h34v5H7v-5z"
              fill="#FFF"
            />
          </svg>
        </span>
        <div
          class="section"
        >
          <strong
            class="title"
          >
            Terra.alert.error
          </strong>
        </div>
      </div>
    </div>
  </div>
  <div>
    <button
      data-testid="show-banner"
      type="button"
    >
      Toggle Notification Banner
    </button>
    <div
      aria-live="polite"
    >
      <div
        class="alert-base error narrow"
        data-terra-application-notification-banner="error"
        role="alert"
      >
        <div
          class="body body-std"
        >
          <span
            class="icon"
          >
            <svg
              aria-hidden="true"
              class="tui-Icon icon IconError"
              focusable="false"
              height="1em"
              viewBox="0 0 48 48"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 0c13.3 0 24 10.7 24 24S37.3 48 24 48 0 37.3 0 24C0 10.8 10.7 0 23.9 0h.1z"
                fill="#E50000"
              />
              <path
                d="M7 21.5h34v5H7v-5z"
                fill="#FFF"
              />
            </svg>
          </span>
          <div
            class="section"
          >
            <strong
              class="title"
            >
              Terra.alert.error
            </strong>
          </div>
        </div>
      </div>
    </div>
    <div>
      <button
        data-testid="show-2nd-banner"
        type="button"
      >
        Toggle Notification Banner
      </button>
    </div>
  </div>
</div>
`;

exports[`useNotificationBanners Nested Notification Banner Provider unregisters banner with top-level Banner Provider 1`] = `
<div>
  <div
    aria-live="polite"
  >
    <div
      class="alert-base error narrow"
      data-terra-application-notification-banner="error"
      role="alert"
    >
      <div
        class="body body-std"
      >
        <span
          class="icon"
        >
          <svg
            aria-hidden="true"
            class="tui-Icon icon IconError"
            focusable="false"
            height="1em"
            viewBox="0 0 48 48"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M24 0c13.3 0 24 10.7 24 24S37.3 48 24 48 0 37.3 0 24C0 10.8 10.7 0 23.9 0h.1z"
              fill="#E50000"
            />
            <path
              d="M7 21.5h34v5H7v-5z"
              fill="#FFF"
            />
          </svg>
        </span>
        <div
          class="section"
        >
          <strong
            class="title"
          >
            Terra.alert.error
          </strong>
        </div>
      </div>
    </div>
  </div>
  <div>
    <button
      data-testid="show-banner"
      type="button"
    >
      Toggle Notification Banner
    </button>
    <div
      aria-live="polite"
    >
      <div
        class="alert-base error narrow"
        data-terra-application-notification-banner="error"
        role="alert"
      >
        <div
          class="body body-std"
        >
          <span
            class="icon"
          >
            <svg
              aria-hidden="true"
              class="tui-Icon icon IconError"
              focusable="false"
              height="1em"
              viewBox="0 0 48 48"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 0c13.3 0 24 10.7 24 24S37.3 48 24 48 0 37.3 0 24C0 10.8 10.7 0 23.9 0h.1z"
                fill="#E50000"
              />
              <path
                d="M7 21.5h34v5H7v-5z"
                fill="#FFF"
              />
            </svg>
          </span>
          <div
            class="section"
          >
            <strong
              class="title"
            >
              Terra.alert.error
            </strong>
          </div>
        </div>
      </div>
    </div>
    <div>
      <button
        data-testid="show-2nd-banner"
        type="button"
      >
        Toggle Notification Banner
      </button>
    </div>
  </div>
</div>
`;

exports[`useNotificationBanners Nested Notification Banner Provider unregisters banner with top-level Banner Provider 2`] = `
<div>
  <div
    aria-live="polite"
  />
  <div>
    <button
      data-testid="show-banner"
      type="button"
    >
      Toggle Notification Banner
    </button>
    <div
      aria-live="polite"
    >
      <div
        class="alert-base error narrow"
        data-terra-application-notification-banner="error"
        role="alert"
      >
        <div
          class="body body-std"
        >
          <span
            class="icon"
          >
            <svg
              aria-hidden="true"
              class="tui-Icon icon IconError"
              focusable="false"
              height="1em"
              viewBox="0 0 48 48"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 0c13.3 0 24 10.7 24 24S37.3 48 24 48 0 37.3 0 24C0 10.8 10.7 0 23.9 0h.1z"
                fill="#E50000"
              />
              <path
                d="M7 21.5h34v5H7v-5z"
                fill="#FFF"
              />
            </svg>
          </span>
          <div
            class="section"
          >
            <strong
              class="title"
            >
              Terra.alert.error
            </strong>
          </div>
        </div>
      </div>
    </div>
    <div>
      <button
        data-testid="show-2nd-banner"
        type="button"
      >
        Toggle Notification Banner
      </button>
    </div>
  </div>
</div>
`;

exports[`useNotificationBanners Nested Notification Banner Provider unregisters banner with top-level Banner Provider 3`] = `
<div>
  <div
    aria-live="polite"
  />
  <div>
    <button
      data-testid="show-banner"
      type="button"
    >
      Toggle Notification Banner
    </button>
    <div
      aria-live="polite"
    />
    <div>
      <button
        data-testid="show-2nd-banner"
        type="button"
      >
        Toggle Notification Banner
      </button>
    </div>
  </div>
</div>
`;

exports[`useNotificationBanners NotificationBannerProvider.registerNotificationBanner registers banner when rendered after mount 1`] = `
<div>
  <div>
    <button
      data-testid="show-banner"
      type="button"
    >
      Toggle Notification Banner
    </button>
  </div>
</div>
`;

exports[`useNotificationBanners NotificationBannerProvider.registerNotificationBanner registers banner when rendered after mount 2`] = `
<div>
  <div
    aria-live="polite"
  >
    <div
      class="alert-base error narrow"
      data-terra-application-notification-banner="error"
      role="alert"
    >
      <div
        class="body body-std"
      >
        <span
          class="icon"
        >
          <svg
            aria-hidden="true"
            class="tui-Icon icon IconError"
            focusable="false"
            height="1em"
            viewBox="0 0 48 48"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M24 0c13.3 0 24 10.7 24 24S37.3 48 24 48 0 37.3 0 24C0 10.8 10.7 0 23.9 0h.1z"
              fill="#E50000"
            />
            <path
              d="M7 21.5h34v5H7v-5z"
              fill="#FFF"
            />
          </svg>
        </span>
        <div
          class="section"
        >
          <strong
            class="title"
          >
            Terra.alert.error
          </strong>
        </div>
      </div>
    </div>
  </div>
  <div>
    <button
      data-testid="show-banner"
      type="button"
    >
      Toggle Notification Banner
    </button>
  </div>
</div>
`;

exports[`useNotificationBanners NotificationBannerProvider.unregisterNotificationBanner unregisters notification banner when id is provided 1`] = `
<div>
  <div
    aria-live="polite"
  >
    <div
      class="alert-base error narrow"
      data-terra-application-notification-banner="error"
      role="alert"
    >
      <div
        class="body body-std"
      >
        <span
          class="icon"
        >
          <svg
            aria-hidden="true"
            class="tui-Icon icon IconError"
            focusable="false"
            height="1em"
            viewBox="0 0 48 48"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M24 0c13.3 0 24 10.7 24 24S37.3 48 24 48 0 37.3 0 24C0 10.8 10.7 0 23.9 0h.1z"
              fill="#E50000"
            />
            <path
              d="M7 21.5h34v5H7v-5z"
              fill="#FFF"
            />
          </svg>
        </span>
        <div
          class="section"
        >
          <strong
            class="title"
          >
            Terra.alert.error
          </strong>
        </div>
      </div>
    </div>
  </div>
  <div>
    <button
      data-testid="show-banner"
      type="button"
    >
      Toggle Notification Banner
    </button>
  </div>
</div>
`;

exports[`useNotificationBanners NotificationBannerProvider.unregisterNotificationBanner unregisters notification banner when id is provided 2`] = `
<div>
  <div
    aria-live="polite"
  />
  <div>
    <button
      data-testid="show-banner"
      type="button"
    >
      Toggle Notification Banner
    </button>
  </div>
</div>
`;

exports[`useNotificationBanners renders with children 1`] = `<div />`;

exports[`useNotificationBanners renders with children 2`] = `
<div>
  <div>
    child
  </div>
</div>
`;

exports[`useNotificationBanners renders with children that has a an banner 1`] = `
<div>
  <div
    aria-live="polite"
  >
    <div
      class="alert-base error narrow"
      data-terra-application-notification-banner="error"
      role="alert"
    >
      <div
        class="body body-std"
      >
        <span
          class="icon"
        >
          <svg
            aria-hidden="true"
            class="tui-Icon icon IconError"
            focusable="false"
            height="1em"
            viewBox="0 0 48 48"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M24 0c13.3 0 24 10.7 24 24S37.3 48 24 48 0 37.3 0 24C0 10.8 10.7 0 23.9 0h.1z"
              fill="#E50000"
            />
            <path
              d="M7 21.5h34v5H7v-5z"
              fill="#FFF"
            />
          </svg>
        </span>
        <div
          class="section"
        >
          <strong
            class="title"
          >
            Terra.alert.error
          </strong>
        </div>
      </div>
    </div>
  </div>
  <div>
    <button
      data-testid="show-banner"
      type="button"
    >
      Toggle Notification Banner
    </button>
  </div>
</div>
`;
